### PR TITLE
feat: implement pre-action step

### PR DESCRIPTION
**Pull request by KnisterPeter**
This PR does implement the pre-action step for remote actions.

### CODE REVIEW

**Comment by KnisterPeter** [comment]
@nektos/act-maintainers With this PR the pre/post steps could be considered done. Please review

**Comment by ChristopherHX** [comment]
Hmm, something went wrong for me.............
```
[w.yml/test] 🚀  Start image=catthehacker/ubuntu:act-latest
[w.yml/test]   🐳  docker pull image=catthehacker/ubuntu:act-latest platform= username= forcePull=false
[w.yml/test]   🐳  docker create image=catthehacker/ubuntu:act-latest platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[w.yml/test]   🐳  docker run image=catthehacker/ubuntu:act-latest platform= entrypoint=["/usr/bin/tail" "-f" "/dev/null"] cmd=[]
[w.yml/test]   🐳  docker exec cmd=[mkdir -m 0777 -p /var/run/act] user=root workdir=
[w.yml/test]   ☁  git clone 'https://ubuntu.fritz.box:3042/christopher/script' # ref=multiplejsfiles
[w.yml/test] ⭐ Run Pre christopher/script@multiplejsfiles
[w.yml/test]   🐳  docker exec cmd=[node /var/run/act/actions/christopher-script@multiplejsfiles/post.js] user= workdir=
| node:internal/modules/cjs/loader:936
|   throw err;
|   ^
|
| Error: Cannot find module '/var/run/act/actions/christopher-script@multiplejsfiles/post.js'
|     at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
|     at Function.Module._load (node:internal/modules/cjs/loader:778:27)
|     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
|     at node:internal/main/run_main_module:17:47 {
|   code: 'MODULE_NOT_FOUND',
|   requireStack: []
| }
[w.yml/test]   ❌  Failure - Pre christopher/script@multiplejsfiles
Error: exit with `FAILURE`: 1
```
@KnisterPeter 
How did you test that? I think there is a missing docker cp. Maybe the action is already cached for you to not require the docker cp, there is a bug in act which doesn't delete all docker volume.

File: pkg/runner/action.go [hunk 1]
```diff
@@ -424,6 +424,74 @@ func getOsSafeRelativePath(s, prefix string) string {
...
+			actionLocation := ""
+			if actionPath != "" {
+				actionLocation = path.Join(actionDir, actionPath)
+			} else {
+				actionLocation = actionDir
+			}
+
+			_, containerActionDir := getContainerActionPaths(stepModel, actionLocation, rc)
+
+			containerArgs := []string{"node", path.Join(containerActionDir, action.Runs.Post)}
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply pkg/runner/action.go:480]: Oh yes :man_facepalming: 
I thought I added a test for that yesterday as well. I'll carefully re-inspect all that code.
The other PRs should be merged before it anyway.

File: pkg/runner/action.go [hunk 2]
```diff
@@ -424,6 +424,74 @@ func getOsSafeRelativePath(s, prefix string) string {
...
+			log.Debugf("executing remote job container: %s", containerArgs)
+
+			return rc.execJobContainer(containerArgs, *step.getEnv(), "", "")(ctx)
+
+		case model.ActionRunsUsingComposite:
+			step.getCompositeRunContext().updateCompositeRunContext(step.getRunContext(), step)
+			return step.getCompositeSteps().post(ctx)
```
Comment (KnisterPeter) [inline pkg/runner/action.go:487]: Also here...

**Review by KnisterPeter — COMMENTED** [review]
It seems the test in `testdata` is not added to the runner tests.

**Comment by ChristopherHX** [comment]
> The other PRs should be merged before it anyway.

Shure, but it is more interesting for me to test the result of all PRs :).

**Comment by KnisterPeter** [comment]
> @KnisterPeter How did you test that? I think there is a missing docker cp. Maybe the action is already cached for you to not require the docker cp, there is a bug in act which doesn't delete all docker volume.
> 

I'm not really understanding this up to now. Locally all the tests are running fine for me. I've clean all docker volumes and made sure there are not pre-cloned actions and also no existing running docker containers.
I'm still investigating it.

**Comment by ChristopherHX** [comment]
Pre actions have missing action.yml defaults? I expect these are populated.
Local composite action have a pre steps? I expect these are skipped.
Otherwise good.

**Comment by KnisterPeter** [comment]
The local action pre-steps need to run if it is a composite action which includes remote actions with pre-steps.

**Comment by ChristopherHX** [comment]
> The local action pre-steps need to run if it is a composite action which includes remote actions with pre-steps.

Only for your implementation? Why? `actions/runner` behaves differently.
You can create an composite actions.yml file in a run step, then run it in the next uses step.

**Comment by KnisterPeter** [comment]
> > The local action pre-steps need to run if it is a composite action which includes remote actions with pre-steps.
> 
> Only for your implementation? Why? `actions/runner` behaves differently. You can create an composite actions.yml file in a run step, then run it in the next uses step.

I will look into that. I thought I have seen this behavior on github, but in fact I cannot confirm that it works like this. Maybe it was good to go on vacation :smile:

**Review by cplee — APPROVED** [review]
looks great!

**Pull request merged**